File: Day1-9/6.hs
import           Data.List.Split
import qualified Data.Set        as Set

part1 :: [String] -> Int
part1 groups = sum $ map length sets
  where
    sets = map (Set.fromList . (filter (/= '\n'))) groups

part2 :: [String] -> Int
part2 groups = sum $ map allAnswered groups
  where
    allAnswered s =
      length $
      foldl1 Set.intersection $ filter (\x -> (length x) > 0) individuals
      where
        individuals = map Set.fromList $ splitOn "\n" s

main = do
  content <- readFile "6.txt"
  let groups = splitOn "\n\n" content
  putStrLn $ show . part1 $ groups
  putStrLn $ show . part2 $ groups

File: Day1-9/6.py
from functools import reduce

if __name__ == "__main__":
    with open("6.txt") as f:
        groups = f.read().split("\n\n")

    part1 = sum(len(set(g.replace("\n", ""))) for g in groups)
    print(f"Part 1: {part1}")

    sets = [[set(s) for s in g.strip().split("\n")] for g in groups]
    part2 = sum(len(reduce(lambda x, y: x & y, s)) for s in sets)
    print(f"Part 2: {part2}")
